src/bin/simple_attachment.rs

```diff
@@ -0,0 +1,175 @@
+///
+/// Example: download a photo by name
+///
+/// > simple_attachment download-photo --name=photo.png
+#[derive(Debug, Parser)]
+struct Cli {
+    #[clap(flatten)]
+    args: Args,
+
+    #[clap(subcommand)]
+    cmd: Cmd,
+}
+
+/// Args needed for any command,
+#[derive(Debug, Parser)]
+struct Args {
+    /// The Ditto App ID to sync with (found at portal.ditto.live)
+    #[clap(long, env = "APP_ID")]
+    app_id: String,
+
+    /// The Playground token used to authenticate (found at portal.ditto.live)
+    #[clap(long, env = "PLAYGROUND_TOKEN")]
+    playground_token: String,
+}
+
+#[derive(Debug, Subcommand)]
+enum Cmd {
+    UploadPhoto {
+        /// Path to a file to upload as an attachment
+        #[clap(long)]
+        path: PathBuf,
+    },
+    DownloadPhoto {
+        /// Name of the attachment file to download
+        #[clap(long)]
+        name: String,
+    },
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    dotenv::dotenv().ok();
+    let cli = Cli::parse();
+
+    // Initialize Ditto SDK client
+    let args = &cli.args;
+    let app_id = AppId::from_str(&args.app_id)?;
+    let ditto = Ditto::builder()
+        .with_root(Arc::new(PersistentRoot::from_current_exe()?))
+        .with_minimum_log_level(LogLevel::Debug)
+        .with_identity(move |ditto_root| {
+            let shared_token = args.playground_token.clone();
+            let enable_cloud_sync = true;
+            let custom_auth_url = None;
+            OnlinePlayground::new(
+                ditto_root,
+                app_id,
+                shared_token,
+                enable_cloud_sync,
+                custom_auth_url,
+            )
+        })?
+        .build()?;
+
+    // Begin sync, then open the Ditto Store so we can insert or query documents
+    ditto.start_sync()?;
+    let store = ditto.store();
+
+    match cli.cmd {
+        Cmd::UploadPhoto { path } => {
+            upload_photo(store, &path).await?;
+        }
+        Cmd::DownloadPhoto { name } => {
+            download_photo(store, &name).await?;
+        }
+    }
+
+    Ok(())
+}
+
+async fn upload_photo(store: &Store, path: &Path) -> Result<()> {
+    let photo_name = path.file_name().and_then(|s| s.to_str()).unwrap_or("photo");
+
+    let photo_attachment = store.new_attachment(path, HashMap::default()).await?;
+    let _result = store
+        .execute(
+            "INSERT INTO COLLECTION photos (photo_attachment ATTACHMENT) DOCUMENTS (:photo_doc)",
+            Some(
+                serde_json::json!({
+                    "photo_doc": {
+                        "photo_name": photo_name,
+                        "photo_attachment": photo_attachment
+                    }
+                })
+                .into(),
+            ),
+        )
+        .await?;
+
+    println!("Uploaded photo with name '{photo_name}'");
+    Ok(())
+}
+
+async fn download_photo(store: &Store, name: &str) -> Result<()> {
+    let result = store
```

Comment:
We should improve on this to make it properly "async"/synchronized (and feel free to adjust documentation so that we're better at educating about this):

```suggestion
async fn download_photo(store: &Store, name: &str) -> Result<()> {
    receive_photo_document(store, name).await?;
    let result = store
```

```rs
async fn receive_photo_document(store: &Store, name: &str) -> Result<()> {
    struct FoundOne;
    let (tx, rx) = some::async_channel();
    let observer = store.register_observer(
        "SELECT * FROM COLLECTION photos (photo_attachment ATTACHMENT)",
        Some(
            serde_json::json!({
                "photo_name": name
            })
            .into(),
        ),
        move |query_result| {
            if query_result.item_count() >= 1 {
                _ = tx.send(FoundOne);
            }
        },
    );
    let FoundOne = rx.await?;
    observer.cancel()?;
    Ok(())
}
```

  - Aside: there is a bug in `StoreObserver`'s `fn is_cancelled()`, and maybe in the docs, I don't believe the handle needs to be held to avoid cancellation, on the contrary, cancellation now may need to be explicit.

Reply:
I had to make a few edits but overall managed to make this pattern work. I had to use a tokio::sync::watch channel because it's the only tokio one that can send without awaiting and is also cloneable (unlike oneshot). I think that attachments needs a bit of an API overhaul in Rust, because if I were an end-user I would be pretty frustrated with this number of hoops to jump through

Reply:
To be fair, this "await doc to arrive" is a rough edge in all of Ditto's API, which is what, in turn:
  - makes `.exec()` kind of a footgun;
  - makes people put excessive logic in the observers.

Reply:
Regarding the channel, https://docs.rs/tokio/1.38.0/tokio/sync/oneshot/struct.Sender.html#method.send is non-blocking, but would require `Option::take()`ing it:

```rs
let mut tx = Some(tx);
// …
move |…| {
    _ = tx.take().map(|tx| tx.send(thing));
}
```